### ISSUE TITLE
HostFXR should service hostpolicy targeted by the app

### ISSUE THREAD

**Issue by gkhanna79** [dotnet/cli#2452]
When HostFXR needs to lookup the serviced version of hostpolicy, it currently looks it up using the version/name info that was baked into it at compile-time, since it is updated in lock-step with hostpolicy.dll.

This will not work when we will need to publish a new hostfxr (which has a different version) that needs to work with existing SharedFX and load a serviced version of hostpolicy corresponding to the FX.

CC @piotrpMSFT @brthor

**Comment by GuardRex** [comment]
@gkhanna79 Does my recent exception working with a test shared framework app pertain to this issue?
```
Error loading hostpolicy 1.0.1-rc2-002296; interface mismatch between hostpolicy [32771] 
and hostfxr [120] Specifically, the structure of corehost_init_t has changed, do not know 
how to interpret it
```

**Comment by schellap** [comment]
Can you try this workaround?
https://github.com/dotnet/cli/issues/2389

**Comment by schellap** [comment]
Could you please try removing 1.0.1-rc2-002296 and target the newer FX.

**Comment by GuardRex** [comment]
@schellap Ok, but that host policy is being pulled in :sparkles: auto-magically :sparkles:
```json
"dependencies": {
    "Microsoft.AspNetCore.Diagnostics": "1.0.0-*",
    "Microsoft.AspNetCore.Server.IISIntegration": "1.0.0-*",
    "Microsoft.Extensions.Logging.Console": "1.0.0-*",
    "Microsoft.AspNetCore.StaticFiles": "1.0.0-*",
    "Microsoft.AspNetCore.Mvc": "1.0.0-*",
    "Microsoft.AspNetCore.Server.Kestrel": "1.0.0-*",
    "Microsoft.NETCore.App": {
        "type": "platform",
        "version": "1.0.0-*"
    }
},
"frameworks": {
    "netcoreapp1.0": {
        "imports": [ "dnxcore50", "portable-net451+win8" ]
    }
},
```
```
.NET Command Line Tools (1.0.0-rc2-002363)

Product Information:
 Version:     1.0.0-rc2-002363
 Commit Sha:  065e8352fc

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.10586
 OS Platform: Windows
 RID:         win10-x64
```
Perhaps, I should wait for `-2392` and see if this issue clears up??

**Comment by schellap** [comment]
```
   "Microsoft.NETCore.App": {
        "type": "platform",
        "version": "1.0.0-rc2-002363"
    }
```

The "version" field should be shared FX that you just installed.
`dir C:\Program Files\dotnet\shared\Microsoft.NETCore.App\`

**Comment by schellap** [comment]
Can you point me to what your runtimeconfig.json contains in the bin drop cotnains?

**Comment by GuardRex** [comment]
@schellap Ok, thanks.

`testshared.runtimeconfig.json`
```json
{
  "runtimeOptions": {
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "1.0.0-rc2-23931"
    }
  }
}
```

**Comment by schellap** [comment]
`dir C:\Program Files\dotnet\shared\Microsoft.NETCore.App\`

or `which dotnet` relative dir?

**Comment by GuardRex** [comment]
![capture](https://cloud.githubusercontent.com/assets/1622880/14481548/bcccc93e-00f9-11e6-9f81-9329961a4e3d.PNG)

**Comment by schellap** [comment]
Can you target 1.0.0-rc2-3002363?

**Comment by GuardRex** [comment]
I can, but I have the Azure VM deprovisioned, so I can't run the same test I ran earlier. I was planning on spinning it back up to test again in the morning. I can post back if you like with the result of the change to the TFM you provided tomrrrow.

**Comment by schellap** [comment]
The runtimeconfig.json will then be updated with this version and the host should pick it up and load the compatible hostpolicy.dll.

**Comment by GuardRex** [comment]
@schellap Is the way I specified `Microsoft.NETCore.App` initially (`-*`) incorrect? ... or is this something that will just not happen shortly with updates?

**Comment by schellap** [comment]
@GuardRex do you have a `global.json` in an ancestor of your working dir? Basically, it seems from the dotnet that got picked up, you have 1.0.0-rc2-002363, but do you have an global.json override?

**Comment by GuardRex** [comment]
No, I'm not using a `global.json` with any of my apps, including this test app.

**Comment by schellap** [comment]
When I did a restore with 1.0.0-*:

I got this error. May be they should make it red in color or something.

error: Unable to resolve 'Microsoft.NETCore.App (>= 1.0.0)' for '.NETCoreApp,Version=v1.0'.

**Comment by GuardRex** [comment]
Not happening here ...
```
log  : Restoring packages for c:\_APPLICATIONS\testshared\project.json...
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/microsoft.aspnetcore.diagnostics/index.json
info :   CACHE https://api.nuget.org/v3-flatcontainer/microsoft.aspnetcore.diagnostics/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/microsoft.aspnetcore.server.iisintegration/index.json
info :   GET https://api.nuget.org/v3-flatcontainer/microsoft.aspnetcore.server.iisintegration/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/microsoft.extensions.logging.console/index.json
info :   CACHE https://api.nuget.org/v3-flatcontainer/microsoft.extensions.logging.console/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/microsoft.aspnetcore.staticfiles/index.json
info :   CACHE https://api.nuget.org/v3-flatcontainer/microsoft.aspnetcore.staticfiles/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/microsoft.aspnetcore.mvc/index.json
info :   CACHE https://api.nuget.org/v3-flatcontainer/microsoft.aspnetcore.mvc/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/system.security.cryptography.algorithms/index.json
info :   CACHE https://api.nuget.org/v3-flatcontainer/system.security.cryptography.algorithms/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/system.security.cryptography.x509certificates/index.json
info :   CACHE https://api.nuget.org/v3-flatcontainer/system.security.cryptography.x509certificates/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/microsoft.aspnetcore.server.kestrel/index.json
info :   CACHE https://api.nuget.org/v3-flatcontainer/microsoft.aspnetcore.server.kestrel/index.json
info :   NotFound https://api.nuget.org/v3-flatcontainer/microsoft.aspnetcore.server.iisintegration/index.json 231ms
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/microsoft.netcore.app/index.json
info :   GET https://api.nuget.org/v3-flatcontainer/microsoft.netcore.app/index.json
info :   NotFound https://api.nuget.org/v3-flatcontainer/microsoft.netcore.app/index.json 242ms
info : Committing restore...
log  : Writing lock file to disk. Path: c:\_APPLICATIONS\testshared\project.lock.json
log  : c:\_APPLICATIONS\testshared\project.json
log  : Restore completed in 2860ms.

NuGet Config files used:
    c:\_APPLICATIONS\testshared\NuGet.Config
    C:\Users\XXXXX\AppData\Roaming\NuGet\NuGet.Config

Feeds used:
    https://www.myget.org/F/aspnetcidev/api/v3/index.json
    https://api.nuget.org/v3/index.json
```

**Comment by GuardRex** [comment]
Why do I have a feeling I found another one??? :imp:

**Comment by schellap** [comment]
`md temp & cd temp`
`dotnet new`
`dotnet restore`

Now change the project.json and do a restore?

**Comment by GuardRex** [comment]
![capture](https://cloud.githubusercontent.com/assets/1622880/14481873/9d850a5c-00fc-11e6-9a93-1e3a61700d2e.PNG)

**Comment by GuardRex** [comment]
The repro for the project is available (for troubleshooting IIS stuff with Chris and Pawel earlier): https://github.com/GuardRex/sharedapp

**Comment by GuardRex** [comment]
oh, sorry  ... I didn't do the last step "change the `project.json`" .. brb

**Comment by schellap** [comment]
Yeah put the wildcard syntax there...

**Comment by GuardRex** [comment]
```json
{
  "version": "1.0.0-*",
  "compilationOptions": {
    "emitEntryPoint": true
  },
  "dependencies": {
    "Microsoft.NETCore.App": {
      "type": "platform",
      "version": "1.0.0-*"
    }
  },
  "frameworks": {
    "netcoreapp1.0": {}
  }
}
```
Result ...
![capture](https://cloud.githubusercontent.com/assets/1622880/14481970/5d893e54-00fd-11e6-9db9-eda11fa1c663.PNG)

**Comment by schellap** [comment]
Not found, that is probably the issue. Can you not specify wildcard and try this out with latest bits?

**Comment by GuardRex** [comment]
I have latest, right? 
```
.NET Command Line Tools (1.0.0-rc2-002363)

Product Information:
 Version:     1.0.0-rc2-002363
 Commit Sha:  065e8352fc

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.10586
 OS Platform: Windows
 RID:         win10-x64
```
The one above at https://github.com/dotnet/cli/issues/2452#issuecomment-209209591 was the one that I ran on the raw `dotnet new` project (no mods to `project.json`. It called for `"version": "1.0.0-rc2-3002363"`

**Comment by schellap** [comment]
I am wondering why your runtimeconfig.json ends up with an older version of the FX. With the runtimeconfig.json containing 1.0.0-rc2-3002363, you are still facing this issue?

**Comment by GuardRex** [comment]
Let me see if I can confirm it just running with Kestrel locally on the `testshared` test app. Be back in a few minutes.

**Comment by GuardRex** [comment]
Ok, so here's what I've got. The `project.json` is ...
```json
"dependencies": {
    "Microsoft.AspNetCore.Diagnostics": "1.0.0-*",
    "Microsoft.AspNetCore.Server.IISIntegration": "1.0.0-*",
    "Microsoft.Extensions.Logging.Console": "1.0.0-*",
    "Microsoft.AspNetCore.StaticFiles": "1.0.0-*",
    "Microsoft.AspNetCore.Mvc": "1.0.0-*",
    "Microsoft.AspNetCore.Server.Kestrel": "1.0.0-*",
    "Microsoft.NETCore.App": {
        "type": "platform",
        "version": "1.0.0-*"
    }
},
"frameworks": {
    "netcoreapp1.0": {
        "imports": [ "dnxcore50", "portable-net451+win8" ]
    }
},
```
The `testshared.runtimeconfig.json` shows this ...
```json
{
  "runtimeOptions": {
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "1.0.0-rc2-23931"
    }
  }
}
```
The restore looks like this ...
```
log  : Restoring packages for c:\_APPLICATIONS\testshared\project.json...
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/microsoft.aspnetcore.diagnostics/index.json
info :   CACHE https://api.nuget.org/v3-flatcontainer/microsoft.aspnetcore.diagnostics/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/microsoft.aspnetcore.server.iisintegration/index.json
info :   GET https://api.nuget.org/v3-flatcontainer/microsoft.aspnetcore.server.iisintegration/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/microsoft.extensions.logging.console/index.json
info :   CACHE https://api.nuget.org/v3-flatcontainer/microsoft.extensions.logging.console/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/microsoft.aspnetcore.staticfiles/index.json
info :   CACHE https://api.nuget.org/v3-flatcontainer/microsoft.aspnetcore.staticfiles/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/microsoft.aspnetcore.mvc/index.json
info :   CACHE https://api.nuget.org/v3-flatcontainer/microsoft.aspnetcore.mvc/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/system.security.cryptography.algorithms/index.json
info :   CACHE https://api.nuget.org/v3-flatcontainer/system.security.cryptography.algorithms/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/system.security.cryptography.x509certificates/index.json
info :   CACHE https://api.nuget.org/v3-flatcontainer/system.security.cryptography.x509certificates/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/microsoft.aspnetcore.server.kestrel/index.json
info :   CACHE https://api.nuget.org/v3-flatcontainer/microsoft.aspnetcore.server.kestrel/index.json
info :   CACHE https://myget-2e16.kxcdn.com/artifacts/aspnetcidev/nuget/v3/flatcontainer/microsoft.netcore.app/index.json
info :   GET https://api.nuget.org/v3-flatcontainer/microsoft.netcore.app/index.json
info :   NotFound https://api.nuget.org/v3-flatcontainer/microsoft.aspnetcore.server.iisintegration/index.json 314ms
info :   NotFound https://api.nuget.org/v3-flatcontainer/microsoft.netcore.app/index.json 280ms
info : Committing restore...
log  : Writing lock file to disk. Path: c:\_APPLICATIONS\testshared\project.lock.json
log  : c:\_APPLICATIONS\testshared\project.json
log  : Restore completed in 2304ms.

NuGet Config files used:
    c:\_APPLICATIONS\testshared\NuGet.Config
    C:\Users\XXXXX\AppData\Roaming\NuGet\NuGet.Config

Feeds used:
    https://www.myget.org/F/aspnetcidev/api/v3/index.json
    https://api.nuget.org/v3/index.json
```
The publish is ...
```
Publishing testshared for .NETCoreApp,Version=v1.0
Compiling testshared for .NETCoreApp,Version=v1.0

Compilation succeeded.
    0 Warning(s)
    0 Error(s)

Time elapsed 00:00:01.8714719
 

publish: Published to c:\_APPLICATIONS\testshared\bin\Release\netcoreapp1.0\publish
Published 1/1 projects successfully
```
... and the exception hits when doing a `dotnet .\testshared.dll` from the `publish` folder.
![capture](https://cloud.githubusercontent.com/assets/1622880/14482210/051b4840-0100-11e6-8e12-4a522f412137.PNG)

**Comment by schellap** [comment]
To get you unblocked, try the no wildcard please in the project.json.
The back compat issue is fixed only in newer version starting 002349.

**Comment by GuardRex** [comment]
Ok ... brb

**Comment by schellap** [comment]
If the wildcard influences what goes inside the project.lock.json then that could be a bug or a question on the usage.

**Comment by GuardRex** [comment]
It warned on this in restore ...
```
warn : Dependency specified was Microsoft.NETCore.App (>= 1.0.0-rc2-3002363) but ended up 
with Microsoft.NETCore.App 1.0.0-rc3-23909.
```
Same on publish ...
```
c:\_APPLICATIONS\testshared\project.json(21,35): warning NU1007: Dependency specified 
was Microsoft.NETCore.App >= 1.0.0-rc2-3002363 but ended up with Microsoft.NETCore.App 
1.0.0-rc3-23909.
```
... and no :heart: on the execution ...
![capture](https://cloud.githubusercontent.com/assets/1622880/14482306/eae089bc-0100-11e6-946e-e5dfba4bb5ab.PNG)

**Comment by schellap** [comment]
Interestingly, I see that if you put the wildcard the version picked is 23931 in the project lock.json.
I think the new CLI needs to get on the ASP.NET feed.

As last resort, another thing you could try out is uninstall (remove or rename to xyz) the 23931 in `Microsoft.NETCore.App\1.0.0-rc2-23931` and to confirm, manually edit the runtimeconfig.json to target the latest framework.

I did the following on your app and it worked fine:
1. Download latest zip file from CLI and ran the app  against it: https://dotnetcli.blob.core.windows.net/dotnet/beta/Binaries/Latest/dotnet-dev-win-x64.latest.zip

{
  "runtimeOptions": {
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "1.0.0-rc2-23931"
    }
  }
}

**Comment by GuardRex** [comment]
Manually editing the `testshared.runtimeconfig.json` and setting it to `"version": "1.0.0-rc2-3002363"` ... :sparkles: works :sparkles:, but yeah, it seems like the larger question is 'What's wrong with the restore (if anything) under these conditions?'

**Comment by schellap** [comment]
You are using a different feed, that feed needs to get the updated packages from CLI.
```
NuGet.config

    <add key="AspNetCI" value="https://www.myget.org/F/aspnetcidev/api/v3/index.json" />

```

**Comment by GuardRex** [comment]
Ok, so this is just a funky lag condition on the feed getting updated.

**Comment by schellap** [comment]
Try this to your NuGet.config and you don't have to muck with the runtimeconfig.json. I don't suggest this as a permanent change, but a way to unblock yourselves.

```
    <add key="NETCore" value="https://www.myget.org/F/dotnet-core/api/v3/index.json" />
```

**Comment by GuardRex** [comment]
mmmm ... I added that one, and I tried to go back to a `1.0.0-*` package ... and ... :boom:

![capture](https://cloud.githubusercontent.com/assets/1622880/14482670/ae3fcae6-0104-11e6-8986-1d5ec9612af4.PNG)

**Comment by schellap** [comment]
no wildcard no.

**Comment by GuardRex** [comment]
oh, ok. .... but this is a known issue (bug), correct? Wildcard there will work down the road, right?

**Comment by schellap** [comment]
In other words, you should target a framework you have on your machine. There  is a conflicting idea of what the wildcard means -- does that mean any on the feed or any on the machine?

**Comment by schellap** [comment]
Although, from the user point of view, just use 1.0.0-* this needs to be thought about, please file an issue.

**Comment by GuardRex** [comment]
ah ... thanks ... that's what I thought. It feels like it's a good design question. I'll open it and link back here in case anyone runs into this. Thanks for your help tonight.

**Comment by schellap** [comment]
Fixed by #2790
